### PR TITLE
docs: add versioning guide

### PR DESCRIPTION
**Pull request by joshblack**
This is a follow-up to: https://github.com/primer/react/pull/2753 with a narrowed scope of changes that I believe we have consensus on with respect to semver 👀 Figured it'd be useful to get this document into `contributor-docs` and add to it over time as we run into things in our release process.

Would love to know what you all think of this format and if there is anything that should be added, edited, or if there is a better approach to documenting these kinds of conventions so that we can refer to them in reviews and in our release process.

### CODE REVIEW

File: contributor-docs/versioning.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+|           | A prop is deprecated                                                | `minor`     |
+|           | A prop is removed                                                   | `major`     |
+
+## Reference
+
+### The type of a prop is broadened
+
+semver bump: **minor**
+
+When a type is broadeded, it is now possible for that type to include both the
```
Comment (lesliecdubs) [inline contributor-docs/versioning.md:59]: > broadeded

Typo?

Comment (joshblack) [reply contributor-docs/versioning.md:59]: Hah, great catch 😂

File: contributor-docs/versioning.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+## Changes
+
+| Category  | Type of change                                                      | semver bump |
+| :-------- | :------------------------------------------------------------------ | :---------- |
+| Component | A component is added                                                | `minor`     |
+|           | A component is deprecated                                           | `minor`     |
+|           | A component is removed                                              | `major`     |
+| Props     | A prop is added                                                     | `minor`     |
+|           | [The type of a prop is broadened](#the-type-of-a-prop-is-broadened) | `minor`     |
+|           | [The type of a prop is narrowed](#the-type-of-a-prop-is-narrowed)   | `major`     |
```
Comment (broccolinisoup) [inline contributor-docs/versioning.md:49]: I love these so much! 💖

**Review by broccolinisoup — APPROVED** [review]
🔥 

How can we make sure this doc becomes a living document as we discover cases/examples in release rotations or in general really? 

Would it be helpful to add this to the release conductor checklist like an optional one maybe but encouraged? 🤔

**Pull request merged**